docs/usage/cryptomanager.md

```diff
@@ -0,0 +1,156 @@
+both singletons.
+
+_Aside:_
+
+> Future work could be done to enable either multiple `CryptoManager` instances
+> with separate NSS DBs, or to enable multiple `JSSProvider` instances to
+> reference one or more `CryptoManager` instances during normal operation.
+> This currently will require significant restructuring as many internal
+> provider methods (such as `Cipher`, `Signature` and others) get the global
+> singleton instance directly (via `CryptoManager.getInstance()`).
+>
+> Additionally, NSS provides the option to load multiple NSS DBs into the
+> current instance, which might satisfy the use case of multiple
+> `CryptoManager` instances as well.
+
+---
+
+However, the existence of `JSSLoader` and subsequently allowing JSS to load
+via the standard provider `java.security` file has complicated this slightly,
+to enable new use cases for JSS.
+
+Take the following code snippet for instance:
+
+```java
+import org.mozilla.jss.CryptoManager;
+
+public class Example {
+  public static void main(String[] args) throws Exception {
+    CryptoManager cm = CryptoManager.getInstance();
+    // Additional code elided.
+  }
+}
+```
+
+When used with JSS loaded via `java.security` override, the developer would
+rightly expect that JSS will load first, allowing `CryptoManager` to return
+a valid instance.
+
+In order to facilitate this however, when the internal `instance` field is
+`NULL`, JSS must first try to load itself via the Provider interface. That is,
+it needs to do something akin to:
+
+```java
+java.security.Provider p = Security.getProvider("Mozilla-JSS");
+```
+
+in order to force Java to attempt to load JSS. Any equivalent call into the
+Java security interface would also suffice (such as
+`Signature.getInstance(algo, "Mozilla-JSS")`), but this call limits the size
+of the resulting object and makes clear the intent.
+
+_Aside:_
+
+> This becomes a touch tricky though. In particular, `CryptoManager.instance`
+> access is usually locked, to prevent modification by one thread while
+> another is reading it. This is locked at the class level (via a
+> `synchronized (CryptoManager.class)` statement explicitly or implicitly
+> in the synchronized `initialize()` method). So, inside `getInstance()`, we
+> explicitly acquire the lock to check the value of `instance`, release it for
+> the provider call -- in case it loads JSS from the provider interface --
+> and then explicitly re-acquire it to return the value. This should help to
+> prevent race conditions and returning a partially-initialized
+> `CryptoManager` instance before it is fully ready.
+
+Otherwise, the call will fail and we'd require code changes, such as:
+
+```java
+import java.security.*;
+
+import org.mozilla.jss.CryptoManager;
+
+public class Example {
+  public static void main(String[] args) throws Exception {
+    Provider p = Security.getProvider("Mozilla-JSS");
+    CryptoManager cm = CryptoManager.getInstance();
```

Comment:
I'm thinking eventually we might do something like this:
```
JSSProvider provider = (JSSProvider) Security.getProvider("Mozilla-JSS");
CryptoManager cm = provider.getCryptoManager();
```
and deprecate/drop `CryptoManager.getInstance()`. Or maybe even replace `CryptoManager` with `JSSProvider`.

Reply:
I think neither of these are quite right. If I were to make these changes (they're big, time-consuming and not immediately clear what the benefit is), I'd probably:

 - Make `JSSProvider` load out of a config file, keep state internally.
 - Make `CryptoManager` an internal implementation detail and have it correlate to exactly one NSS DB. It will back-end the `KeyStoreSpi` and `TrustStoreSpi` code. Instances will be singleton-per-NSS-DB and stored/cached on the associated `JSSProvider` instance for use throughout the JCA.
 - Drop all non-JCA interfaces when duplicated by JCA interfaces.
 - Make JCA interfaces both compatible with SunJSSE/... and more compliant. I'd like to see keys &c interop when possible and other improvements.

But, who knows. :-)

Reply:
Most of the magic also occurs in the "more compliant" and "compatible" steps -- which the rest isn't necessary for and could be done now. The benefit of the restructuring is that it would let us do the .p12/jks stuff, which isn't as likely to work well otherwise as you can't control where keys come from as easily. So I think it is unlikely this restructuring is likely to occur now.

Reply:
+1 to all of the above. There might be no major immediate benefit, but I think whenever we make modifications we should do it with these goals in mind. That way we won't accumulate a huge technical debt when we have an actual requirement that depends on this in the future.